src-electron/generator/helper-zcl.js

```diff
@@ -730,24 +731,27 @@ function zcl_global_commands(options) {
  * @param {*} options
  * @returns Promise of attribute iteration.
  */
-function zcl_attributes(options) {
+async function zcl_attributes(options) {
   // If used at the toplevel, 'this' is the toplevel context object.
   // when used at the cluster level, 'this' is a cluster
-  let promise = templateUtil
-    .ensureZclPackageIds(this)
-    .then((packageIds) => {
-      if ('id' in this) {
-        // We're functioning inside a nested context with an id, so we will only query for this cluster.
-        return queryZcl.selectAttributesByClusterIdIncludingGlobal(
-          this.global.db,
-          this.id,
-          packageIds
-        )
-      } else {
-        return queryZcl.selectAllAttributes(this.global.db, packageIds)
-      }
-    })
-    .then((atts) => templateUtil.collectBlocks(atts, options, this))
+  let packageIds = await templateUtil.ensureZclPackageIds(this)
+  let attributes = ''
+  if ('id' in this) {
+    // We're functioning inside a nested context with an id, so we will only query for this cluster.
+    attributes = await queryZcl.selectAttributesByClusterIdIncludingGlobal(
+      this.global.db,
+      this.id,
+      packageIds
+    )
+    attributes = await upgrade.computeStoragePolicyForGlobalAttributes(
```

Comment:
This doesn't feel right, layer-wise. The point of `upgrade.js` is to perform upgrades when you READ IN the .zap files.
You are here using `upgrade.js` when you're generating stuff.

Not saying that it's the wrong thing, functionally, but I'd then go and move the logic out of `upgrade.js`, put it somewhere else, and maybe have both places call the common place.

Reply:
will make follow up PR, thanks